File: android/salom_ai/lib/core/api/api_models.dart
import 'package:json_annotation/json_annotation.dart';

// -- Chat --

class ChatOut {
  final String reply;
  final int conversationId;

  ChatOut({required this.reply, required this.conversationId});

  factory ChatOut.fromJson(Map<String, dynamic> json) {
    return ChatOut(
      reply: json['reply'] as String,
      conversationId: json['conversation_id'] as int,
    );
  }
}

class ChatStreamEvent {
  final String type;
  final String? content;
  final int? conversationId;
  final String? message;

  ChatStreamEvent({required this.type, this.content, this.conversationId, this.message});

  factory ChatStreamEvent.fromJson(Map<String, dynamic> json) {
    return ChatStreamEvent(
      type: json['type'] as String,
      content: json['content'] as String?,
      conversationId: json['conversation_id'] as int?,
      message: json['message'] as String?,
    );
  }
}

class TokenPair {
  final String accessToken;
  final String refreshToken;

  TokenPair({required this.accessToken, required this.refreshToken});

  factory TokenPair.fromJson(Map<String, dynamic> json) {
    return TokenPair(
      accessToken: json['access_token'] as String,
      refreshToken: json['refresh_token'] as String,
    );
  }
}

enum MessageRole { user, assistant, system }

class SearchResult {
  final String? title;
  final String? url;
  final String? date;

  SearchResult({this.title, this.url, this.date});

  factory SearchResult.fromJson(Map<String, dynamic> json) {
    return SearchResult(
      title: json['title'] as String?,
      url: json['url'] as String?,
      date: json['date'] as String?,
    );
  }
}

class MessageDTO {
  final int id;
  final MessageRole role;
  final String? text;
  final DateTime? createdAt;
  final List<String>? imageUrls;
  final List<SearchResult>? searchResults;
  final String? perplexityModel;
  final List<String>? fileUrls;
  final String? audioUrl;

  MessageDTO({
    required this.id,
    required this.role,
    this.text,
    this.createdAt,
    this.imageUrls,
    this.searchResults,
    this.perplexityModel,
    this.fileUrls,
    this.audioUrl,
  });

  factory MessageDTO.fromJson(Map<String, dynamic> json) {
    return MessageDTO(
      id: json['id'] as int,
      role: MessageRole.values.firstWhere((e) => e.toString().split('.').last == json['role'], orElse: () => MessageRole.user),
      text: json['text'] as String?,
      createdAt: json['created_at'] != null ? DateTime.parse(json['created_at']) : null,
      imageUrls: (json['image_urls'] as List?)?.map((e) => e as String).toList(),
      searchResults: (json['search_results'] as List?)?.map((e) => SearchResult.fromJson(e)).toList(),
      perplexityModel: json['perplexity_model'] as String?,
      fileUrls: (json['file_urls'] as List?)?.map((e) => e as String).toList(),
      audioUrl: json['audio_url'] as String?,
    );
  }
}

class ConversationSummary {
  final int id;
  final String? title;
  final DateTime? updatedAt;
  final int? messageCount;

  ConversationSummary({required this.id, this.title, this.updatedAt, this.messageCount});

  factory ConversationSummary.fromJson(Map<String, dynamic> json) {
    return ConversationSummary(
      id: json['id'] as int,
      title: json['title'] as String?,
      updatedAt: json['updated_at'] != null ? DateTime.parse(json['updated_at']) : null,
      messageCount: json['message_count'] as int?,
    );
  }
}

class ConversationListResponse {
  final List<ConversationSummary> conversations;
  final int? total;

  ConversationListResponse({required this.conversations, this.total});

  factory ConversationListResponse.fromJson(dynamic json) {
    if (json is List) {
      return ConversationListResponse(
        conversations: json.map((e) => ConversationSummary.fromJson(e)).toList(),
        total: json.length,
      );
    }

    final map = json as Map<String, dynamic>;
    return ConversationListResponse(
      conversations: (map['conversations'] as List)
          .map((e) => ConversationSummary.fromJson(e))
          .toList(),
      total: map['total'] as int?,
    );
  }
}

class ConversationMessagesResponse {
  final int conversationId;
  final List<MessageDTO> messages;
  final int total;

  ConversationMessagesResponse({required this.conversationId, required this.messages, required this.total});

  factory ConversationMessagesResponse.fromJson(Map<String, dynamic> json) {
    return ConversationMessagesResponse(
      conversationId: json['conversation_id'] ?? 0,
      messages: (json['messages'] as List?)?.map((e) => MessageDTO.fromJson(e)).toList() ?? [],
      total: json['total'] as int? ?? 0,
    );
  }
}

class OAuthUser {
  final int id;
  final String? email;
  final String? displayName;
  final String? authProvider;
  final DateTime? createdAt;

  OAuthUser({required this.id, this.email, this.displayName, this.authProvider, this.createdAt});

  factory OAuthUser.fromJson(Map<String, dynamic> json) {
    return OAuthUser(
      id: json['id'] as int,
      email: json['email'] as String?,
      displayName: json['display_name'] as String?,
      authProvider: json['auth_provider'] as String?,
      createdAt: json['created_at'] != null ? DateTime.parse(json['created_at']) : null,
    );
  }
}

class StatusMessageResponse {
  final String? detail;
  StatusMessageResponse({this.detail});
  factory StatusMessageResponse.fromJson(Map<String, dynamic> json) => StatusMessageResponse(detail: json['detail'] as String?);
}

class SubscriptionPlan {
  final String code;
  final String name;
  final int priceUzs;
  final int? monthlyMessages;
  final int? monthlyTokens;
  final List<Map<String, String>>? benefits;

  SubscriptionPlan({
    required this.code,
    required this.name,
    required this.priceUzs,
    this.monthlyMessages,
    this.monthlyTokens,
    this.benefits,
  });

  factory SubscriptionPlan.fromJson(Map<String, dynamic> json) {
    return SubscriptionPlan(
      code: json['code'] as String,
      name: json['name'] as String,
      priceUzs: json['price_uzs'] as int,
      monthlyMessages: json['monthly_messages'] as int?,
      monthlyTokens: json['monthly_tokens'] as int?,
      benefits: (json['benefits'] as List?)
          ?.map((e) => Map<String, String>.from(e as Map))
          .toList(),
    );
  }
}

class CurrentSubscriptionResponse {
  final String? plan;
  final bool active;
  final DateTime? startedAt;
  final DateTime? expiresAt;

  CurrentSubscriptionResponse({this.plan, required this.active, this.startedAt, this.expiresAt});

  factory CurrentSubscriptionResponse.fromJson(Map<String, dynamic> json) {
    return CurrentSubscriptionResponse(
      plan: json['plan'] as String?,
      active: json['active'] as bool,
      startedAt: json['started_at'] != null ? DateTime.parse(json['started_at']) : null,
      expiresAt: json['expires_at'] != null ? DateTime.parse(json['expires_at']) : null,
    );
  }
}

class SubscribeResponse {
  final String? provider;
  final int? amountUzs;
  final String? status;
  final String? checkoutUrl;

  SubscribeResponse({this.provider, this.amountUzs, this.status, this.checkoutUrl});

  factory SubscribeResponse.fromJson(Map<String, dynamic> json) {
    return SubscribeResponse(
      provider: json['provider'] as String?,
      amountUzs: json['amount_uzs'] as int?,
      status: json['status'] as String?,
      checkoutUrl: json['checkout_url'] as String?,
    );
  }
}

class FeedbackResponse {
  final int id;
  final String content;

  FeedbackResponse({required this.id, required this.content});

  factory FeedbackResponse.fromJson(Map<String, dynamic> json) {
    return FeedbackResponse(
      id: json['id'] as int,
      content: json['content'] as String,
    );
  }
}

// -- AI Models --

class AIModel {
  final String id;
  final String name;
  final String? tier;
  final bool vision;
  final int? limit;
  final int? usage;

  AIModel({
    required this.id,
    required this.name,
    this.tier,
    this.vision = false,
    this.limit,
    this.usage,
  });

  factory AIModel.fromJson(Map<String, dynamic> json) {
    return AIModel(
      id: json['id'] as String? ?? json['model_id'] as String? ?? '',
      name: json['name'] as String? ?? json['display_name'] as String? ?? '',
      tier: json['tier'] as String?,
      vision: json['vision'] as bool? ?? json['supports_vision'] as bool? ?? false,
      limit: json['limit'] as int? ?? json['daily_limit'] as int?,
      usage: json['usage'] as int? ?? json['daily_usage'] as int?,
    );
  }
}

// -- Search --

class MessageSearchHit {
  final int conversationId;
  final String? conversationTitle;
  final int messageId;
  final String? text;
  final String? role;

  MessageSearchHit({
    required this.conversationId,
    this.conversationTitle,
    required this.messageId,
    this.text,
    this.role,
  });

  factory MessageSearchHit.fromJson(Map<String, dynamic> json) {
    return MessageSearchHit(
      conversationId: json['conversation_id'] as int,
      conversationTitle: json['conversation_title'] as String?,
      messageId: json['message_id'] as int? ?? json['id'] as int? ?? 0,
      text: json['text'] as String?,
      role: json['role'] as String?,
    );
  }
}

class SearchResponse {
  final List<MessageSearchHit> results;
  final int total;

  SearchResponse({required this.results, this.total = 0});

  factory SearchResponse.fromJson(dynamic json) {
    if (json is List) {
      return SearchResponse(
        results: json.map((e) => MessageSearchHit.fromJson(e)).toList(),
        total: json.length,
      );
    }
    final map = json as Map<String, dynamic>;
    return SearchResponse(
      results: (map['results'] as List?)?.map((e) => MessageSearchHit.fromJson(e)).toList() ?? [],
      total: map['total'] as int? ?? 0,
    );
  }
}

// -- File Upload --

class FileUploadResponse {
  final String url;
  final String? fileId;
  final String? filename;

  FileUploadResponse({required this.url, this.fileId, this.filename});

  factory FileUploadResponse.fromJson(Map<String, dynamic> json) {
    return FileUploadResponse(
      url: json['url'] as String? ?? json['file_url'] as String? ?? '',
      fileId: json['file_id'] as String? ?? json['id']?.toString(),
      filename: json['filename'] as String? ?? json['name'] as String?,
    );
  }
}

// -- Image Generation --

class ImageGenerationResponse {
  final String? imageUrl;
  final String? status;
  final String? error;

  ImageGenerationResponse({this.imageUrl, this.status, this.error});

  factory ImageGenerationResponse.fromJson(Map<String, dynamic> json) {
    return ImageGenerationResponse(
      imageUrl: json['image_url'] as String? ?? json['url'] as String?,
      status: json['status'] as String?,
      error: json['error'] as String?,
    );
  }
}

// -- Usage Stats --

class UsageStatsResponse {
  final PlanInfo? plan;
  final UsageLimits? limits;
  final UsageData? usage;

  UsageStatsResponse({this.plan, this.limits, this.usage});

  factory UsageStatsResponse.fromJson(Map<String, dynamic> json) {
    return UsageStatsResponse(
      plan: json['plan'] != null ? PlanInfo.fromJson(json['plan']) : null,
      limits: json['limits'] != null ? UsageLimits.fromJson(json['limits']) : null,
      usage: json['usage'] != null ? UsageData.fromJson(json['usage']) : null,
    );
  }
}

class PlanInfo {
  final String name;
  final String code;
  final bool isPro;

  PlanInfo({required this.name, required this.code, this.isPro = false});

  factory PlanInfo.fromJson(Map<String, dynamic> json) {
    return PlanInfo(
      name: json['name'] as String? ?? '',
      code: json['code'] as String? ?? '',
      isPro: json['is_pro'] as bool? ?? false,
    );
  }
}

class UsageLimits {
  final int? messages;
  final int? images;
  final int? files;
  final int? voiceMinutes;

  UsageLimits({this.messages, this.images, this.files, this.voiceMinutes});

  factory UsageLimits.fromJson(Map<String, dynamic> json) {
    return UsageLimits(
      messages: json['messages'] as int?,
      images: json['images'] as int?,
      files: json['files'] as int?,
      voiceMinutes: json['voice_minutes'] as int?,
    );
  }
}

class UsageData {
  final int? messagesUsed;
  final int? imagesUsed;
  final int? filesUsed;
  final int? voiceMinutesUsed;
  final Map<String, int>? perModel;

  UsageData({this.messagesUsed, this.imagesUsed, this.filesUsed, this.voiceMinutesUsed, this.perModel});

  factory UsageData.fromJson(Map<String, dynamic> json) {
    return UsageData(
      messagesUsed: json['messages_used'] as int? ?? json['messages'] as int?,
      imagesUsed: json['images_used'] as int? ?? json['images'] as int?,
      filesUsed: json['files_used'] as int? ?? json['files'] as int?,
      voiceMinutesUsed: json['voice_minutes_used'] as int? ?? json['voice_minutes'] as int?,
      perModel: (json['per_model'] as Map?)?.map((k, v) => MapEntry(k as String, v as int)),
    );
  }
}

// -- Voice STT --

class STTResponse {
  final String? text;
  final String? language;

  STTResponse({this.text, this.language});

  factory STTResponse.fromJson(Map<String, dynamic> json) {
    return STTResponse(
      text: json['text'] as String?,
      language: json['language'] as String?,
    );
  }
}

File: Salom-Ai-iOS/Salom-Ai-iOS/Managers/APIModels.swift
//
//  APIModels.swift
//  Salom-Ai-iOS
//
//  Created by Codex on 23/01/25.
//

import Foundation

enum MessageRole: String, Codable {
    case user
    case assistant
    case system
}

struct TokenPair: Codable {
    let accessToken: String
    let refreshToken: String
    let tokenType: String
}

// Telegram "login via bot code" — response of /auth/telegram/code/start.
// (keyDecodingStrategy = .convertFromSnakeCase maps bot_url/expires_in.)
struct TelegramCodeStartResponse: Codable {
    let token: String
    let botUrl: String
    let expiresIn: Int
}

struct OAuthUser: Codable {
    let id: Int
    let email: String?
    let displayName: String?
    let avatarUrl: String?
    let authProvider: String?
    let createdAt: Date?

    enum CodingKeys: String, CodingKey {
        case id
        case email
        case displayName = "display_name"
        case avatarUrl = "avatar_url"
        case authProvider = "auth_provider"
        case createdAt = "created_at"
    }
}

struct StatusMessageResponse: Decodable {
    let detail: String?

    private enum CodingKeys: String, CodingKey { case detail }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        if let message = try? container.decode(String.self, forKey: .detail) {
            detail = message
        } else if let failure = try? container.decode(CardChargeFailure.self, forKey: .detail),
                  let data = try? JSONEncoder().encode(failure) {
            // Keep APIError's public shape stable while preserving the structured
            // decline payload for the subscription funnel.
            detail = String(data: data, encoding: .utf8)
        } else {
            detail = nil
        }
    }
}

struct ChatReplyResponse: Codable {
    let reply: String
    let conversationId: Int
    // Note: No custom CodingKeys needed - decoder uses .convertFromSnakeCase
}
/// A web source returned with a search-grounded answer.
struct Citation: Codable, Hashable {
    let title: String
    let url: String
}

struct ChatStreamEvent: Codable {
    let type: String
    let content: String?
    let conversationId: Int?
    let message: String?
    // Web-search extras: stage=="searching_web"/"generating_image" on a "status"
    // event, citations on a "citations" event, upgrade==true on limit events.
    let stage: String?
    let citations: [Citation]?
    let upgrade: Bool?
    // Auto-generated image URL on an "image" event.
    let url: String?
}

struct ChatSavePayload: Codable {
    let conversationId: Int
    let userText: String
    let assistantText: String
    
    enum CodingKeys: String, CodingKey {
        case conversationId = "conversation_id"
        case userText = "user_text"
        case assistantText = "assistant_text"
    }
}

struct SearchResult: Codable, Identifiable {
    let id = UUID()
    let title: String?
    let url: String?
    let date: String?
    
    enum CodingKeys: String, CodingKey {
        case title
        case url
        case date
    }
}

struct MessageDTO: Codable, Identifiable {
    let id: Int
    let role: MessageRole
    let text: String?
    let createdAt: Date?
    let searchResults: [SearchResult]?
    let perplexityModel: String?
    let imageUrls: [String]?
    let fileUrls: [String]?
    let audioUrl: String?
}

struct ConversationSummary: Codable, Identifiable {
    let id: Int
    let title: String?
    let updatedAt: Date?
    let messageCount: Int?
    
    enum CodingKeys: String, CodingKey {
        case id
        case title
        case updatedAt = "updated_at"
        case messageCount = "message_count"
    }
}

struct ConversationListResponse: Codable {
    let conversations: [ConversationSummary]
    let total: Int?
}

struct ConversationDetailResponse: Codable {
    let id: Int
    let title: String?
    let autoTitled: Bool
    let createdAt: Date
    let updatedAt: Date
    let messageCount: Int
    
    enum CodingKeys: String, CodingKey {
        case id
        case title
        case autoTitled = "auto_titled"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case messageCount = "message_count"
    }
}

struct ConversationMessagesResponse: Codable {
    let conversationId: Int
    let messages: [MessageDTO]
    let total: Int
}

//struct ConversationMessageDTO: Codable, Identifiable {
//    let id: Int
//    let role: MessageRole
//    let text: String?
//    let createdAt: Date?
//    let searchResults: [SearchResult]?
//    let perplexityModel: String?
//    let imageUrls: [String]?
//    let fileUrls: [String]?
//    let audioUrl: String?
//    
//    enum CodingKeys: String, CodingKey {
//        case id
//        case role
//        case text
//        case createdAt = "created_at"
//        case searchResults = "search_results"
//        case perplexityModel = "perplexity_model"
//        case imageUrls = "image_urls"
//        case fileUrls = "file_urls"
//        case audioUrl = "audio_url"
//    }
//}

struct MessageSearchHit: Codable, Identifiable {
    let id: Int
    let conversationId: Int?
    let role: MessageRole?
    let text: String?
    let createdAt: Date?
    let conversationTitle: String?
    
    enum CodingKeys: String, CodingKey {
        case id
        case conversationId = "conversation_id"
        case role
        case text
        case createdAt = "created_at"
        case conversationTitle = "conversation_title"
    }
}

struct SearchResponse: Decodable {
    let results: [MessageSearchHit]
    let total: Int?
    
    init(results: [MessageSearchHit], total: Int? = nil) {
        self.results = results
        self.total = total
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        if let decodedResults = try container.decodeIfPresent([MessageSearchHit].self, forKey: .results) {
            results = decodedResults
        } else if let messages = try container.decodeIfPresent([MessageSearchHit].self, forKey: .messages) {
            results = messages
        } else {
            results = []
        }
        total = try container.decodeIfPresent(Int.self, forKey: .total)
    }
    
    enum CodingKeys: String, CodingKey {
        case results
        case messages
        case total
    }
}

struct PerplexityUsageResponse: Codable {
    let requestsUsed: Int?
    let requestsLimit: Int?
    let deepResearchUsed: Int?
    let remainingRequests: Int?
}

struct PerplexityModelInfo: Codable, Identifiable {
    var id: String { model }
    let model: String
    let name: String?
    let description: String?
    let inputPricePer1k: Double?
    let outputPricePer1k: Double?
    let useCase: String?
    let note: String?
    

}

struct PerplexityChatResponse: Codable {
    let reply: String
    let conversationId: Int
    let searchResults: [SearchResult]?
    let modelUsed: String?
    

}

struct PerplexityResearchResponse: Codable {
    let report: String
    let searchResults: [SearchResult]?
    let usageInfo: UsageInfo?
    

}

struct UsageInfo: Codable {
    let inputTokens: Int?
    let outputTokens: Int?
    let costUsd: Double?
    let reasoningEffort: String?
    

}

struct SpeechToTextResponse: Codable {
    let text: String
    let durationSec: Double?
    let costUsd: Double?
    

}

typealias STTResponse = SpeechToTextResponse

struct VoiceChatHeaders {
    let transcript: String?
    let reply: String?
    let conversationId: Int?
    let costUsd: Double?
}

struct SubscriptionPlan: Codable, Identifiable {
    var id: String { code }
    let code: String
    let name: String
    let priceUzs: Int
    let durationDays: Int?
    let monthlyMessages: Int?
    let monthlyTokens: Int?
    let benefits: [[String: String]]?

    /// Cost per day, derived from priceUzs / durationDays. Never hardcodes 30 —
    /// reads what the admin configured. Falls back to 30 only if API omits it.
    var pricePerDay: Double {
        let days = max(1, durationDays ?? 30)
        return Double(priceUzs) / Double(days)
    }

    /// Is this plan paid (not the free tier)?
    var isPaid: Bool { priceUzs > 0 }

    /// Localized benefit string for a given UI language. Falls back through uz → en.
    func benefit(at index: Int, lang: String) -> String? {
        guard let benefits, index < benefits.count else { return nil }
        let row = benefits[index]
        let short = String(lang.prefix(2)).lowercased()
        return row[short] ?? row["uz"] ?? row["en"] ?? row.values.first
    }
}

// MARK: - Win-back / recovery offer (GET /subscriptions/recovery-offer)

/// Returned only for users who started a payment but never finished and have no
/// active sub. Mirrors the web win-back logic so iOS recovers abandoned payments.
struct RecoveryOfferResponse: Codable {
    let eligible: Bool
    let askSurvey: Bool?
    let offers: [RecoveryOffer]?
}

struct RecoveryOffer: Codable, Identifiable {
    var id: String { promoCode }
    let baseCode: String
    let baseName: String
    let basePrice: Int
    let promoCode: String
    let promoPrice: Int
    let discountPct: Int
    let benefits: [[String: String]]?

    /// Localized benefit string, uz → en fallback (same contract as SubscriptionPlan).
    func benefit(at index: Int, lang: String) -> String? {
        guard let benefits, index < benefits.count else { return nil }
        let row = benefits[index]
        let short = String(lang.prefix(2)).lowercased()
        return row[short] ?? row["uz"] ?? row["en"] ?? row.values.first
    }
}

struct CurrentSubscriptionResponse: Codable {
    let plan: String?
    let active: Bool
    let startedAt: Date?
    let expiresAt: Date?
    

}

struct SubscribeResponse: Codable {
    let provider: String?
    let amountUzs: Int?
    let status: String?
    let checkoutUrl: String?
    let paymentId: Int?
}

/// Status of a single payment transaction (`GET /subscriptions/payments/{id}`).
/// Mirrors the web `PaymentResult` polling source — the authoritative per-payment
/// truth used to decide the post-checkout toast (never inferred from `isPro`).
struct PaymentStatusResponse: Codable {
    let id: Int
    let status: String   // waiting_user | pending | pending_callback | paid | failed | cancelled
    let plan: String?
    let provider: String?
}

struct SettingsPayload: Codable {
    let systemPrompt: String?
    let preferences: [String: String]?
    
    enum CodingKeys: String, CodingKey {
        case systemPrompt = "system_prompt"
        case preferences
    }
}

struct UsageStatsResponse: Codable {
    let plan: PlanInfo
    let limits: UsageLimits
    let usage: UsageData
    let resetDate: String?
    

    
    struct PlanInfo: Codable {
        let code: String
        let name: String
        let priceUzs: Int?
        

    }
    
    struct UsageLimits: Codable {
        let fastMessages: Int
        let smartMessages: Int
        let superSmartMessages: Int
        let imageGeneration: Int
        let fileAnalysis: Int
        let voiceMinutes: Int
        let liveQueries: Int
        
        enum CodingKeys: String, CodingKey {
            case fastMessages
            case smartMessages
            case superSmartMessages
            case imageGeneration
            case fileAnalysis
            case voiceMinutes
            case liveQueries
        }
        
        init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            fastMessages = try container.decodeIfPresent(Int.self, forKey: .fastMessages) ?? 0
            smartMessages = try container.decodeIfPresent(Int.self, forKey: .smartMessages) ?? 0
            superSmartMessages = try container.decodeIfPresent(Int.self, forKey: .superSmartMessages) ?? 0
            imageGeneration = try container.decodeIfPresent(Int.self, forKey: .imageGeneration) ?? 0
            fileAnalysis = try container.decodeIfPresent(Int.self, forKey: .fileAnalysis) ?? 0
            voiceMinutes = try container.decodeIfPresent(Int.self, forKey: .voiceMinutes) ?? 0
            liveQueries = try container.decodeIfPresent(Int.self, forKey: .liveQueries) ?? 0
        }
        
        init(fastMessages: Int = 0, smartMessages: Int = 0, superSmartMessages: Int = 0, imageGeneration: Int = 0, fileAnalysis: Int = 0, voiceMinutes: Int = 0, liveQueries: Int = 0) {
            self.fastMessages = fastMessages
            self.smartMessages = smartMessages
            self.superSmartMessages = superSmartMessages
            self.imageGeneration = imageGeneration
            self.fileAnalysis = fileAnalysis
            self.voiceMinutes = voiceMinutes
            self.liveQueries = liveQueries
        }
    }
    
    struct UsageData: Codable {
        let fastMessages: Int
        let smartMessages: Int
        let superSmartMessages: Int
        let imageGeneration: Int
        let fileAnalysis: Int
        let voiceMinutes: Int
        let liveQueries: Int
        
        enum CodingKeys: String, CodingKey {
            case fastMessages
            case smartMessages
            case superSmartMessages
            case imageGeneration
            case fileAnalysis
            case voiceMinutes
            case liveQueries
        }
        
        init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            fastMessages = try container.decodeIfPresent(Int.self, forKey: .fastMessages) ?? 0
            smartMessages = try container.decodeIfPresent(Int.self, forKey: .smartMessages) ?? 0
            superSmartMessages = try container.decodeIfPresent(Int.self, forKey: .superSmartMessages) ?? 0
            imageGeneration = try container.decodeIfPresent(Int.self, forKey: .imageGeneration) ?? 0
            fileAnalysis = try container.decodeIfPresent(Int.self, forKey: .fileAnalysis) ?? 0
            voiceMinutes = try container.decodeIfPresent(Int.self, forKey: .voiceMinutes) ?? 0
            liveQueries = try container.decodeIfPresent(Int.self, forKey: .liveQueries) ?? 0
        }
        
        init(fastMessages: Int = 0, smartMessages: Int = 0, superSmartMessages: Int = 0, imageGeneration: Int = 0, fileAnalysis: Int = 0, voiceMinutes: Int = 0, liveQueries: Int = 0) {
            self.fastMessages = fastMessages
            self.smartMessages = smartMessages
            self.superSmartMessages = superSmartMessages
            self.imageGeneration = imageGeneration
            self.fileAnalysis = fileAnalysis
            self.voiceMinutes = voiceMinutes
            self.liveQueries = liveQueries
        }
    }
    

}

struct FileUploadResponse: Codable {
    let url: String
    let filename: String
}

struct DocGenResponse: Codable {
    let url: String
    let filename: String
}

// MARK: - Card Tokenization

struct TokenizeRequestResponse: Codable {
    let requestId: String
    let phoneHint: String
}

struct TokenizeVerifyResponse: Codable {
    let success: Bool
    let subscription: TokenizeSubscriptionInfo?
    let savedCard: TokenizeSavedCard?

    struct TokenizeSubscriptionInfo: Codable {
        let plan: String
        let active: Bool
        let expiresAt: String
        let autoRenew: Bool
    }

    struct TokenizeSavedCard: Codable {
        let id: Int
        let maskedNumber: String
        let phoneHint: String
    }
}

struct CardChargeFailure: Codable {
    let code: String
    let paymentId: Int?
    let planCode: String
    let amountUzs: Int?
    let canRetry: Bool?

    enum CodingKeys: String, CodingKey {
        case code
        case paymentId = "payment_id"
        case planCode = "plan_code"
        case amountUzs = "amount_uzs"
        case canRetry = "can_retry"
    }
}

struct SavedCard: Codable, Identifiable {
    let id: Int
    let maskedNumber: String
    let phoneHint: String
    let createdAt: Date?
}

struct AutoRenewResponse: Codable {
    let ok: Bool
    let autoRenew: Bool
    let savedCardId: Int?
}

struct CancelSubscriptionResponse: Codable {
    let ok: Bool
    let message: String?
    let expiresAt: String?
}

struct CurrentSubscriptionFull: Codable {
    let plan: String?
    let active: Bool
    let startedAt: Date?
    let expiresAt: Date?
    let autoRenew: Bool?
    let savedCard: SavedCard?
    // Payment recovery: a failed auto-renew paused the sub; the saved card can be retried.
    let inRecovery: Bool?
    let recoveryUntil: Date?
    let canRetry: Bool?
}

struct RetryPaymentResponse: Codable {
    let ok: Bool
    let alreadyActive: Bool?
    let expiresAt: String?
}
